log/zap/field.go

```diff
@@ -59,3 +61,156 @@ func (t trace) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+}
+
+func (s spanContext) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	ctx := jaeger.SpanContext(s)
+	enc.AddString("trace", ctx.TraceID().String())
+	enc.AddString("span", ctx.SpanID().String())
+	enc.AddString("parent", ctx.ParentID().String())
+	enc.AddBool("debug", ctx.IsDebug())
+	enc.AddBool("sampled", ctx.IsSampled())
+	enc.AddBool("firehose", ctx.IsFirehose())
+	enc.AddArray("baggage", baggageKVs(ctx))
+	return nil
+}
+
+type referencedContext jaeger.SpanContext
+
+func (s referencedContext) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	ctx := jaeger.SpanContext(s)
+	enc.AddString("span", ctx.SpanID().String())
+	enc.AddString("parent", ctx.ParentID().String())
+	return nil
+}
+
+type baggageKV struct {
+	key   string
+	value string
+}
+
+func (b baggageKV) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddString("key", b.key)
+	enc.AddString("value", b.value)
+	return nil
+}
+
+type baggageKVs jaeger.SpanContext
+
+func (b baggageKVs) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	jaeger.SpanContext(b).
+		ForeachBaggageItem(func(k, v string) bool {
+			enc.AppendObject(baggageKV{
+				key:   k,
+				value: v,
+			})
+			return true
+		})
+	return nil
+}
+
+type logRecord opentracing.LogRecord
+
+func (l logRecord) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddTime("ts", l.Timestamp)
+	enc.AddArray("fields", logFields(l.Fields))
+	return nil
+}
+
+type logRecords []opentracing.LogRecord
+
+func (l logRecords) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	for _, record := range l {
+		enc.AppendObject(logRecord(record))
+	}
+	return nil
+}
+
+type logField log.Field
+
+func (l logField) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	lf := log.Field(l)
+	enc.AddString("key", lf.Key())
+	enc.AddString("value", fmt.Sprint(lf.Value()))
+	return nil
+}
+
+type logFields []log.Field
+
+func (l logFields) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	for _, field := range l {
+		enc.AppendObject(logField(field))
+	}
+	return nil
+}
+
+type tags opentracing.Tags
+
+func (t tags) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	for k, v := range t {
+		enc.AddString("key", k)
+		enc.AddReflected("value", v)
+	}
+	return nil
+}
+
+type reference opentracing.SpanReference
+
+func (r reference) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	if r.Type == opentracing.ChildOfRef {
+		enc.AddString("type", "child_of")
+	} else if r.Type == opentracing.FollowsFromRef {
+		enc.AddString("type", "follows_from")
+	} else {
+		enc.AddString("type", "unknown")
+	}
+
+	if jCtx, ok := r.ReferencedContext.(jaeger.SpanContext); ok {
+		enc.AddObject("referenced_context", referencedContext(jCtx))
+	}
+
+	return nil
+}
+
+type references []opentracing.SpanReference
+
+func (r references) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	for _, spanReference := range r {
+		enc.AppendObject(reference(spanReference))
+	}
+	return nil
+}
+
+type span struct {
+	span opentracing.Span
+}
```

Comment:
suggest keeping consistent with the rest and use `type span opentracing.Span`

Also might consider type-checking for `*jaeger.Span` in the following function rather than in MarshalLogObject

Reply:
`opentracing.Span` is an interface. We can typedef it, but go doesn't allow implementing methods on interfaces

Reply:
The approach of declaring a new type, and then converting it to use the old type seems pretty contrived to me. What are the benefits of doing this, as opposed to a plain or embedded struct?

Reply:
New structs cause memory allocations as they are converted to `interface{}` when passing through Zap API. stdlib also usually uses type aliases instead of creating wrappers, so you could say it's more idiomatic.

Reply:
ty for the explanation!